spec/formatting.md

```diff
@@ -18,3 +18,57 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided to the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is invalid.
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+- **Unresolved variable errors** are a sub-category of resolution errors,
+  and occur when a variable reference cannot be resolved.
+- **Formatting errors** are thrown by expression handlers,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error must also be separately provided.
+
+When an error occurs in the syntax or resolution of an Expression or MarkupStart Option,
+the Expression or MarkupStart in question is processed as if the option was not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector may only match the catch-all VariantKey `*`.
+
```

Comment:
Is this always true? Or is it only true if the selector cannot otherwise select a value. For example, if one of the values of the selector uses an unrecognized keyword, but the evaluated values already matches a separate keyword?

That is, I'm reading the phrase "may only" like it should be "must only". Perhaps you mean "... the selector should (only) match the catch-all..."?

Reply:
```suggestion
When an error occurs within a Selector,
the selector must not match any VariantKey other than the "catch-all" `*`.
```

Given that we don't have the selection language locked down yet, perhaps reversing the statement like this might make it clearer?

Reply:
This is better normative language, but leaves open what happens if (for example) there is no catch-all.

We should probably have a name for the catch-all, such as "default"

I'm not sure I agree with the intent, since it does not allow the selector to do anything except select `*`, even when the error is otherwise recoverable. The example I gave in our call was a plural selector with an unselected typo, e.g.:

```
match {$myValue :number}
when 0 {Nope, no wildebeest here}
when one {You have {$myValue} wildebeest}
when fwe {You have {$myValue} wildebeest} // "few" is misspelled
when many {You have {$myValue} many wildebeest}
when * {My you have a lot of wildebeest!}
```

There's an argument to be made that this selector is toast since you can't be sure that the selection will do the right thing.

Perhaps what we should do (at least for now) is say something like:

> When an error occurs within a Selector, the Selector *must* return a valid message. Selectors *may* define specific error handling behavior. Returning the default value (`*`) is *recommended*

Side note: we should start to be careful with our 2119 keyword usage, because these become conformance criteria.

Reply:
It sounds like this is really about the dual role played by the `*` variant, as it's covering a specific case as well as being a fallback. To cover that concern, we do still allow for e.g. `other` to still be used as a keyword. Altering a bit your example above to do that:

```
match {$myValue :number}
when 0 {Nope, no wildebeest here}
when one {You have {$myValue} wildebeest}
when fwe {You have {$myValue} wildebeest} // "few" is misspelled
when many {You have {$myValue} many wildebeest}
when other {My you have a lot of wildebeest!}
when * {Your wildebeest count: {$myValue}}
```

In most cases, that's probably overkill, but we do allow for it in the spec by falling back to `*` if we don't know any better.

I would presume that that message should raise a validation error during some build step, but during formatting/runtime it would not be considered an error for any numeric `$myValue`.

Reply:
My first reaction is that it would be a parse error, but actually, that would require MessageFormatter to understand what the selector expects/produces. `fwe` is syntactically valid and a different selector (say the old `SelectFormat`) might match/produce a result against the "typo" that plural doesn't.

I'm not wild about having both `other` and `*`. We're making this complicated to write for developers and harder to understand for human translators. Your illustration of the difference is very good, though.

Reply:
> we do still allow for e.g. other

I don't think this is allowed.

The keys are either strings that are matched "as is" (like `select` in MF1), or special keywords that are function specific. And those would be defined in the function registry.

So `other` is allowed, but it does not have any special meaning (like catch all)

I would expect that the registration of `plural` and `ordinal` will allow for `zero`, `one`, `two`, `few`, `many`. But not `other`, the `*` will play that role.

Reply:
> I would expect that the registration of plural and ordinal will allow for zero, one, two, few, many. But not other, the * will play that role.

In some Slavic languages `other` is used specifically for fractions, and thus is not suitable to be used as a generic fallback, which AFAIUI should be grammatically correct for any value of the number.

Reply:
`other` is overloaded for those languages--it is the fallback, but it was also assigned a grammatical purpose (as a workaround). In those languages, one of the defined keywords other than `other` captures all of the integer cases. In the current Intl.PluralRules and Java/ICU implementation, if the pattern isn't localized into the language (or the localizers forget a keyword), the `other` rule captures values that would match `few`, `many` and so forth.

We could allow separate generic (`*`) fallback and a defined keyword fallback (`other`) at the cost of having to modify existing implementation. Would developers and translators understand what to do?

Reply:
This sounds like a discussion that belongs more appropriately in #322 rather than here.

Reply:
Resolving, as the original source has been updated and the rest of this conversation is now moved here, along with @macchiati's reply that got separated from this thread: https://github.com/unicode-org/message-format-wg/discussions/322#discussioncomment-4763944